### DISCUSSION TITLE
Add less intrusive "Warn Before Quitting" option like Google Chrome

### DISCUSSION THREAD

**Discussion by lopopolo** [ghostty-org/ghostty#4009] [Feature Requests, Ideas]
Because ghostty has native tabs, I am not running tmux. As such, I do not want a stray cmd+Q to quit Ghostty and lose all of my shell state. However, the warn dialog has several issues:

- Only fires if a command is running in the shell
- Requires mouse interaction on the dialog (tab and space do not work to select the buttons)
- Prevents shutdown (for example if macOS wants to reboot for updates)

I'd prefer to have it set up so that:

- cmd+q+q unconditionally quits the terminal and terminates all shells and their subprocesses
- cmd+q puts up a temporary UI affordance to indicate how to quit (not a modal!)
- ghostty never blocks shutdown

**Comment by 00-kat** [comment]
Tab, space and arrow keys appear to work on Linux with the GTK dialogue, FWIW.

Although it would be nice if something like Alt+Y and Alt+N worked (don't know what equivalent macOS bindings would be), or maybe just Y and N.

(Similarly with the clipboard one, although that's probably not relevant for this discussion.  And also, if you use a tiling window manager with focus follows mouse enabled, those dialogues appear as separate windows so moving your mouse removes focus from them.  Not sure if this is a GTK issue or a Ghostty issue.)

**Comment by SimenB** [comment]
I don't have to use the mouse to interact with the modal, I can use `return` to quit, or `escape` to abort. `tab` also works to jump between them, so seems odd it doesn't work for you 🤔 

---

Forcing an "Are you sure" on `cmd+q` like Chrome would be very much appreciated, tho!

<img width="315" alt="image" src="https://github.com/user-attachments/assets/bd119e08-d607-4627-9699-d532d0536958" />

**Comment by austin-weeks** [comment]
This feature would be very much appreciated!

**Comment by dkmar** [comment]
Another option is to require it twice. This works for me:
```
keybind = cmd+q>cmd+q=quit
```

**Comment by sandstrom** [comment]
Adding to this topic:

When macOS auto-restarts due to OS updates, it often hangs on the Terminal, because the default behavior of macOS built-in Terminal is to halt if there are running processes.

I would have preferred if macOS terminal could be configured to always proceed with shutdown (for example SIGTERM + SIGKILL after ~30 seconds).

The problem this causes is that often I want OS updates during the night, and when I open up in the morning it's still stuck and I have to sit through 20 min of system updates.

**Reply by orrc** [reply]
Same here. I use tmux (along with the resurrect plugin), so I'm fine with my terminal being killed without warning.

**Reply by orrc** [reply]
It looks like [this config option](https://ghostty.org/docs/config/reference#confirm-close-surface) enables quitting Ghostty without warning:
```
confirm-close-surface = false
```

**Reply by sandstrom** [reply]
Nice!

There is also `window-save-state` for saving the state.
